### PR TITLE
[#3837] improvement(CI): Optimize job of CI

### PR DESCRIPTION
**Pull request by mchades**
### What changes were proposed in this pull request?

 - use gradle cache
 - reuse intermediate compilation results
 - remove unnecessary command

### Why are the changes needed?

Fix: #3837 

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

CI passed

### CODE REVIEW

File: dev/ci/util_free_space.sh [hunk 1]
```diff
@@ -23,12 +23,6 @@ set -eux
 
 if [ "${GITHUB_ACTIONS}" = "true" ]; then
   df -h
-  echo "::group::/usr/local/*"
-  du -hsc /usr/local/*
-  echo "::endgroup::"
-  echo "::group::/usr/local/bin/*"
-  du -hsc /usr/local/bin/*
-  echo "::endgroup::"
```
Comment (yuqi1129) [inline dev/ci/util_free_space.sh:31]: Are these steps very time-consuming?

Comment (mchades) [reply dev/ci/util_free_space.sh:31]: I saw the
I once saw that the free_space step took several minutes to complete, which should have been caused by IO generated by the `du` command. After removing these unnecessary `du` commands, the free_space step only takes a few seconds.

Comment (yuqi1129) [reply dev/ci/util_free_space.sh:31]: Got it.

**Review by yuqi1129 — APPROVED** [review]
LGTM

**Pull request merged**